07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+    
+
+
+{-
+Exercise 2 The first annotation to try out is one for fast indexing into a JoinList. The idea is to cache the size (number of data ele- ments) of each subtree. This can then be used at each step to deter- mine if the desired index is in the left or the right branch.
+We have provided the Sized module that defines the Size type, which is simply a newtype wrapper around an Int. In order to make Sizes more accessible, we have also defined the Sized type class which provides a method for obtaining a Size from a value.
+Use the Sized type class to write the following functions.
+-}
+--Exercise 2
+{-
+Implement the function
+indexJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> Maybe a
+indexJ finds the JoinList element at the specified index. If the index is out of bounds, the function returns Nothing. By an index in a JoinList we mean the index in the list that it represents. That is, consider a safe list indexing function
+(!!?) :: [a] -> Int -> Maybe a [] !!? _ = Nothing
+!!? i | i < 0 = Nothing (x:xs) !!? 0 =Justx (x:xs) !!? i = xs !!? (i-1)
+which returns Just the ith element in a list (starting at zero) if such an element exists, or Nothing otherwise. We also consider an updated function for converting join-lists into lists, just like jlbToList but ignoring the monoidal annotations:
+jlToList :: JoinList m a -> [a]
+jlToList Empty = []
+jlToList (Single _ a) = [a]
+jlToList (Append _ l1 l2) = jlToList l1 ++ jlToList l2
+We can now specify the desired behavior of indexJ. For any index i and join-list jl, it should be the case that
+   (indexJ i jl) == (jlToList jl !!? i)
+That is, calling indexJ on a join-list is the same as first convert- ing the join-list to a list and then indexing into the list. The point, of course, is that indexJ can be more efficient (O(log n) versus O(n), assuming a balanced join-list), because it gets to use the size annotations to throw away whole parts of the tree at once, whereas the list indexing operation has to walk over every element.
+-}
+indexJ :: (Sized b, Monoid b) => Int -> JoinList b a -> Maybe a
+indexJ n (Single _ a)
+    | n == 0    = Just a
+    | otherwise = Nothing
+indexJ n (Append m jl1 jl2)    
+    | n < jl1Size = indexJ n             jl1
+    | otherwise   = indexJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+indexJ _ _ = Nothing
+
+
+
+
+{-
+Implement the function
+dropJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The dropJ function drops the first n elements from a JoinList. This is analogous to the standard drop function on lists. Formally, dropJ should behave in such a way that
+jlToList (dropJ n jl) == drop n (jlToList jl)
+-}
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ n (Single m a)
+    | n >= 1    = Empty
+    | otherwise = Single m a
+dropJ n (Append _ jl1 jl2)
+    | n < jl1Size = dropJ n             jl1 +++ jl2
+    | otherwise   = dropJ (n - jl1Size) jl2
```

Comment:
This implementation is most probably correct. However, the issue is that you will always have to traverse the left joint list until you reach the `Single` constructor. Can you think of a more efficient implementation that considers 3 cases (instead of 2)?

Reply:
@oishikg I was thinking that I can check if `n`, the number of `JoinList`s I am dropping, is greater than the size of the current `JoinList`. If it is, I can return `Empty` straightaway.

Actually, will an implementation with 4 cases be even more efficient? In addition to the extra case mentioned above, I can add a 4th case where I check if `n` is lesser or equals than `0` and if true, I return the current `JoinList` straightaway?

Reply:
@oishikg Ahh I see it now. I think I can introduce a check for when `n == 0`, and if that were true, to return the entire `JoinList` immediately.